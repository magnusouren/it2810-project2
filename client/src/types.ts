export interface User {
  name: string;
  id: string;
  loginState: boolean;
  ratings?: Rating[];
  watchlist?: Movie[];
}

/**
 * The Movie type is based on the API response from
 * https://developers.themoviedb.org/3/movies/get-popular-movies
 * The API response is a bit more complex, but we only need these fields.
 */
export interface Movie {
  adult: boolean;
  backdrop_path: string;
  genre_ids: number[];
  _id: number;
  original_language: string;
  original_title: string;
  overview: string;
  popularity: number;
  poster_path: string;
  release_date: string;
  title: string;
  video: boolean;
  vote_average: number;
  vote_count: number;
}

/**
 * The Category type is based on the API response from
 * https://developers.themoviedb.org/3/genres/get-movie-list
 */
export interface Category {
  id: number;
  name: string;
}

export interface Rating {
  movieId: number;
  rating: number;
}

export type AlphabeticalSort = 'a-z' | 'z-a' | '';

export type RatingSort = 'h-l' | 'l-h' | '';
