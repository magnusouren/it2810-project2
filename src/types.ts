export type User = {
  name: string;
  movieList?: Movie[];
};

export type Movie = {
  id: number;
  title: string;
  year: number;
  rating: number;
  poster: string;
};
