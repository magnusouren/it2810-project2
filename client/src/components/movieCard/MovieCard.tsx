import { Card, CardContent, CardMedia, Typography } from '@mui/material';
import { FC } from 'react';
import { Link } from 'react-router-dom';

import { useUser } from '../../context/UserContext';
import { Movie } from '../../types';
import WatchlistButton from '../watchlistButton/WatchlistButton';
import styles from './MovieCard.module.scss';

interface MovieProps {
  movie: Movie;
}

/**
 *
 * MovieCard component.
 * Takes values from parameters and displays them in an organized card.
 *
 * @param movie
 * @returns {React.JSX.Element}
 */

export const MovieCard: FC<MovieProps> = ({ movie }) => {
  const { user } = useUser();

  return (
    <Card className={styles.card} data-testid='movie-card'>
      <Link to={`/movie/${movie._id}`} state={movie} className={styles.link}>
        <CardMedia
          component='img'
          image={`https://image.tmdb.org/t/p/w400${movie.poster_path}`}
          alt={movie.title + ' poster.'}
          className='movie-card-image'
          loading='lazy'
        />
        <CardContent>
          <Typography gutterBottom component='div' aria-label={`title: ${movie.title}.`}>
            {movie.title}
          </Typography>
          <Typography
            sx={{ color: 'var(--color-text-secondary)' }}
            variant='body2'
            aria-label={`genres:${movie.genre_ids.map((genre) => genre.name).join(', ')}`}
          >
            {movie.genre_ids.map((genre) => genre.name).join(', ')}
          </Typography>
        </CardContent>
      </Link>
      {user && <WatchlistButton movie={movie} user={user} />}
    </Card>
  );
};
