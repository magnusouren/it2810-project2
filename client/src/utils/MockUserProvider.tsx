import { FC, ReactNode, useState } from 'react';
import { vi } from 'vitest';

// Importer UserContext
import { UserContext } from '../context/UserContext'; // Importer UserContext
import { User } from '../types';

interface MockUserProviderProps {
  children: ReactNode;
  mockUser?: User;
  loginMock?: () => void;
  logoutMock?: () => void;
  deleteUserMock?: () => void;
  toggleDarkModeMock?: () => void;
}

const MockUserProvider: FC<MockUserProviderProps> = ({
  children,
  mockUser,
  loginMock = vi.fn(),
  logoutMock = vi.fn(),
  deleteUserMock = vi.fn(),
  toggleDarkModeMock = vi.fn(),
}) => {
  const [user, setUser] = useState<User | undefined>(mockUser);

  const login = loginMock;
  const logout = logoutMock;
  const deleteUser = deleteUserMock;
  const toggleDarkMode = toggleDarkModeMock;

  const value = {
    user,
    setUser,
    login,
    logout,
    deleteUser,
    darkMode: false,
    toggleDarkMode,
  };

  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
};

export default MockUserProvider;
