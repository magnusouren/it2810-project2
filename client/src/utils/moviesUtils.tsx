import { GET_MOVIES, GET_MOVIES_BY_GENRE, GET_MOVIES_BY_RATING, GET_MOVIES_BY_TITLE_AZ } from '../graphql/queries';

export const determineQueryAndVariables = (page: number, genre: string, sort: string, userId: string) => {
  const baseVariables = { page: page, userId: userId };
  const alphabeticalSort = sort === 'a-z' || sort === 'z-a';
  const ratingSort = sort === 'h-l' || sort === 'l-h';

  if (genre) {
    if (alphabeticalSort) {
      return {
        query: GET_MOVIES_BY_TITLE_AZ,
        variables: { ...baseVariables, genreId: genre, order: sort },
      };
    }
    if (ratingSort) {
      return {
        query: GET_MOVIES_BY_RATING,
        variables: { ...baseVariables, genreId: genre, order: sort },
      };
    }
    return {
      query: GET_MOVIES_BY_GENRE,
      variables: { ...baseVariables, genreId: genre },
    };
  }

  if (alphabeticalSort) {
    return {
      query: GET_MOVIES_BY_TITLE_AZ,
      variables: { ...baseVariables, order: sort },
    };
  }
  if (ratingSort) {
    return {
      query: GET_MOVIES_BY_RATING,
      variables: { ...baseVariables, order: sort },
    };
  }
  return {
    query: GET_MOVIES,
    variables: { ...baseVariables },
  };
};
