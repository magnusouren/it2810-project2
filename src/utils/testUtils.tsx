import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { UserProvider } from '../context/UserContext';

// Function to render a component wrapped inside a browserrouter for testing purposes
export const renderWithRouterAndUserContext = (component: JSX.Element) =>
  render(
    <UserProvider>
      <BrowserRouter>{component}</BrowserRouter>
    </UserProvider>,
  );

const queryClient = new QueryClient();

// Function to render a component wrapped inside a browserrouter and queryclient for testing purposes
export const renderWithRouterAndQueryClient = (component: JSX.Element) =>
  render(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>{component}</BrowserRouter>
    </QueryClientProvider>,
  );
