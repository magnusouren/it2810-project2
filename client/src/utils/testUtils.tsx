import { ApolloClient, ApolloProvider, HttpLink, InMemoryCache } from '@apollo/client';
import { MockedProvider } from '@apollo/client/testing';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { User } from '../types';
import MockUserProvider from './MockUserProvider';

const link = new HttpLink({
  uri: 'http://localhost:4000', // Endepunktet til din GraphQL server
});

const client = new ApolloClient({
  cache: new InMemoryCache(),
  link: link,
});

type ProviderMocks = {
  loginMock?: () => void;
  logoutMock?: () => void;
  deleteUserMock?: () => void;
  toggleDarkModeMock?: () => void;
};

interface ProviderProps {
  child: React.ReactNode;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  mocks?: any[];
  mockUser?: User;
  providerMocks?: ProviderMocks;
}

// Function to render a component wrapped inside a browserrouter for testing purposes
export const renderWithProviders = ({ child, mockUser, mocks, providerMocks }: ProviderProps) =>
  render(
    <MockUserProvider mockUser={mockUser} {...providerMocks}>
      <ApolloProvider client={client}>
        <MockedProvider mocks={mocks} addTypename={false}>
          <BrowserRouter>{child}</BrowserRouter>
        </MockedProvider>
      </ApolloProvider>
    </MockUserProvider>,
  );
