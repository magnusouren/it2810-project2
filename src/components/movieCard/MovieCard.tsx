import { Card, CardContent, CardMedia, Typography } from '@mui/material';
import { FC } from 'react';
import { Link } from 'react-router-dom';

import { useUser } from '../../context/UserContext';
import { Movie } from '../../types';
import { getCategoryById } from '../../utils/categoryUtils';
import WatchlistButton from '../watchlistButton/WatchlistButton';
import styles from './MovieCard.module.scss';

interface MovieProps {
  movie: Movie;
}

export const MovieCard: FC<MovieProps> = ({ movie }) => {
  const userContext = useUser();
  if (!userContext) return null;
  const { user, existInWatchlist, toggleMovieInWatchlist } = userContext;

  return (
    <Card sx={{ maxWidth: 200 }} className={styles.card}>
      <Link to={`/movie/${movie.id}`} state={movie} className={styles.link}>
        <CardMedia component='img' image={`https://image.tmdb.org/t/p/w200${movie.poster_path}`} alt={movie.title} />
        <CardContent>
          <Typography gutterBottom component='div'>
            {movie.title}
          </Typography>
          <Typography variant='body2' color='text.secondary'>
            {movie.genre_ids.map((genre) => getCategoryById(genre)?.name).join(', ')}
          </Typography>
        </CardContent>
      </Link>
      {user && (
        <WatchlistButton
          movie={movie}
          existInWatchlist={existInWatchlist}
          toggleMovieInWatchlist={toggleMovieInWatchlist}
          data-testid='watchlist-button'
        />
      )}
    </Card>
  );
};
