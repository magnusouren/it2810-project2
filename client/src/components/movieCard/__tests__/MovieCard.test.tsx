import { screen, waitFor } from '@testing-library/react';

import { Movie, User } from '../../../types';
import { renderWithProviders } from '../../../utils/testUtils';
import { MovieCard } from '../MovieCard';

const mockUser: User = {
  id: '1',
  name: 'Foo Bar',
  loginState: true,
};

describe('MovieCard', () => {
  const movie: Movie = {
    adult: false,
    backdrop_path: '/mockBackdropURL.jpg',
    genre_ids: [
      { _id: 12, name: 'Adventure' },
      { _id: 28, name: 'Action' },
    ], // Adventure, Action
    _id: 111111,
    original_language: 'en',
    original_title: 'Foo Bar',
    overview: 'Simple mock overview',
    popularity: 1000,
    poster_path: '/mockPosterURL.jpg',
    release_date: '1970-01-01',
    title: 'Foo Bar',
    video: false,
    vote_average: 5,
    vote_count: 555,
  };

  it('should match snapshot', async () => {
    const { container } = renderWithProviders({ child: <MovieCard movie={movie} />, mockUser: mockUser });
    await waitFor(() => {
      expect(screen.getByTestId('watchlist-toggle-button')).toBeDefined();
    });
    expect(container).toMatchSnapshot();
  });

  it('should render the watchlist button when user is defined', async () => {
    renderWithProviders({ child: <MovieCard movie={movie} />, mockUser: mockUser });
    await waitFor(() => {
      expect(screen.getByTestId('watchlist-toggle-button')).toBeDefined();
    });
  });

  it('should not render the watchlist button when user is undefined', async () => {
    renderWithProviders({ child: <MovieCard movie={movie} /> });
    expect(screen.queryByTestId('watchlist-toggle-button')).toBeNull();
  });

  it('should render the movie title', () => {
    renderWithProviders({ child: <MovieCard movie={movie} />, mockUser: mockUser });

    const titleElement = screen.getByText(movie.title);
    expect(titleElement).toBeDefined();
  });

  it('should render the movie genres', () => {
    renderWithProviders({ child: <MovieCard movie={movie} />, mockUser: mockUser });

    const genreElement = screen.getByText('Adventure, Action');
    expect(genreElement).toBeDefined();
  });

  it('should render the movie poster', () => {
    renderWithProviders({ child: <MovieCard movie={movie} />, mockUser: mockUser });

    const posterElement = screen.getByAltText(movie.title + ' poster.');
    expect(posterElement).toBeDefined();
    expect(posterElement).toHaveProperty('src', `https://image.tmdb.org/t/p/w400${movie.poster_path}`);
  });
});
