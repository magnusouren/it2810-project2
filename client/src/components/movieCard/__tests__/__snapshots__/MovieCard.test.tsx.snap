// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MovieCard > should match snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root _card_8de2c1 css-bhp9pd-MuiPaper-root-MuiCard-root"
    data-testid="movie-card"
  >
    <a
      class="_link_8de2c1"
      href="/movie/111111"
    >
      <img
        alt="Foo Bar poster."
        class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img movie-card-image css-o69gx8-MuiCardMedia-root"
        loading="lazy"
        src="https://image.tmdb.org/t/p/w400/mockPosterURL.jpg"
      />
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <div
          aria-label="title: Foo Bar."
          class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
        >
          Foo Bar
        </div>
        <p
          aria-label="genres:Adventure, Action"
          class="MuiTypography-root MuiTypography-body2 css-xx3zdw-MuiTypography-root"
        >
          Adventure, Action
        </p>
      </div>
    </a>
    <div
      class="_smallContainer_5dd437"
      data-testid="watchlist-button"
    >
      <button
        aria-label="add movie to watchlist"
        aria-pressed="false"
        class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-sizeSmall MuiToggleButton-primary css-1dlyelz-MuiButtonBase-root-MuiToggleButton-root"
        data-testid="watchlist-toggle-button"
        tabindex="0"
        type="button"
        value="check"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium _add_5dd437 css-i4bv87-MuiSvgIcon-root"
          data-testid="BookmarkAddIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M21 7h-2v2h-2V7h-2V5h2V3h2v2h2v2zm-2 14-7-3-7 3V5c0-1.1.9-2 2-2h7c-.63.84-1 1.87-1 3 0 2.76 2.24 5 5 5 .34 0 .68-.03 1-.1V21z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
